core/drivers/crypto/caam/caam_key.c

```diff
@@ -0,0 +1,746 @@
+			return CAAM_FAILURE;
+
+		*size = alloc + sizeof(key_serialized);
+
+	}
+
+	return CAAM_NO_ERROR;
+}
+
+enum caam_status caam_key_deserialize_from_bn(const struct bignum *inkey,
+					      struct caamkey *outkey,
+					      size_t size_sec)
+{
+	enum caam_status status = CAAM_FAILURE;
+	uint8_t *buf = NULL;
+	size_t size = 0;
+
+	assert(inkey && outkey);
+
+	KEY_TRACE("Deserialization bignum");
+
+	/* Get bignum size */
+	size = crypto_bignum_num_bytes((struct bignum *)inkey);
+
+	/* Allocate temporary buffer */
+	buf = caam_calloc(size);
+	if (!buf)
+		return CAAM_OUT_MEMORY;
+
+	/* Convert bignum to binary */
+	crypto_bignum_bn2bin(inkey, buf);
+
+	status = caam_key_deserialize_from_bin(buf, size, outkey, size_sec);
+
+	caam_key_dump("Output key", outkey);
+
+	caam_free(buf);
+
+	return status;
+}
+
+enum caam_status caam_key_serialize_to_bn(struct bignum *outkey,
+					  const struct caamkey *inkey)
+{
+	enum caam_status status = CAAM_FAILURE;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *buf = NULL;
+	size_t size = 0;
+
+	assert(inkey && outkey);
+
+	KEY_TRACE("Serialization bignum");
+	caam_key_dump("Input key", inkey);
+
+	status = caam_key_serialized_size(inkey, &size);
+	if (status)
+		return status;
+
+	buf = caam_calloc(size);
+	if (!buf)
+		return CAAM_OUT_MEMORY;
+
+	status = caam_key_serialize_to_bin(buf, size, inkey);
+	if (status)
+		goto out;
+
+	res = crypto_bignum_bin2bn(buf, size, outkey);
+	if (res)
+		status = CAAM_FAILURE;
+out:
+	caam_free(buf);
+
+	return status;
+}
+
+#define MAX_DESC_ENTRIES 22
+enum caam_status caam_key_black_encapsulation(struct caamkey *key,
+					      enum caam_key_type key_type)
+{
+	enum caam_status status = CAAM_FAILURE;
+	struct caambuf input_buf = { };
+	struct caam_jobctx jobctx = { };
+	uint32_t *desc = NULL;
+
+	assert(key);
+	assert(!key->is_blob && key->key_type == CAAM_KEY_PLAIN_TEXT);
+	assert(key_type != CAAM_KEY_PLAIN_TEXT);
+
+	KEY_TRACE("Black key encapsulation");
+
+	/* Copy input plain text key to temp buffer */
+	status = caam_calloc_align_buf(&input_buf, key->buf.length);
+	if (status)
+		return status;
+
+	memcpy(input_buf.data, key->buf.data, key->buf.length);
+	cache_operation(TEE_CACHEFLUSH, input_buf.data, input_buf.length);
+
+	/* Re-allocate the output key for black format */
+	caam_key_free(key);
+	key->key_type = key_type;
+
+	status = caam_key_alloc(key);
+	if (status)
+		goto out;
+
+	/* Allocate the descriptor */
+	desc = caam_calloc_desc(MAX_DESC_ENTRIES);
+	if (!desc) {
+		KEY_TRACE("Allocation descriptor error");
+		status = CAAM_OUT_MEMORY;
+		goto out;
+	}
+
+	caam_key_dump("Input key", key);
+
+	caam_desc_init(desc);
+	caam_desc_add_word(desc, DESC_HEADER(0));
+	caam_desc_add_word(desc, LD_KEY(CLASS_1, PKHA_E, key->sec_size));
+	caam_desc_add_ptr(desc, input_buf.paddr);
+
+	if (key_type == CAAM_KEY_BLACK_ECB)
+		caam_desc_add_word(desc, FIFO_ST(CLASS_NO, PKHA_E_AES_ECB_JKEK,
+						 key->sec_size));
+	else if (key_type == CAAM_KEY_BLACK_CCM)
```

Comment:
Add nothing if `key_type` isn't `CAAM_KEY_BLACK_CCM` either?

Reply:
If the key isn't black, no encapsulation is needed. There is an `assert()` at the beginning of the `caam_key_black_encapsulation()` function.

Reply:
Please skip the "if (key_type == CAAM_KEY_BLACK_CCM)" part and only use an `else`.

If the answer is the same for all the other similar comments I've made, do the same there.

Reply:
Ok for this one.

Reply:
When you're testing for as many different values as in the last case I'd prefer a switch-case construct, with a default for error handling.

At this location, I'd prefer just
``` C
if (key_type == CAAM_KEY_BLACK_ECB)
         ...
else
         ...
```
Please try to avoid testing for the same thing twice in the same function since it then in the second test "can't happen" if it was an error condition.

Reply:
Ok, I will check other places where `key_type` is tested